File: reta3/tables.cpp
#include "tables.hpp"
#include <utility>

// Output implementation
Output::Output(const std::string& txt) : txt_(txt) {}

void Output::set_tables(Tables* tables) {
    tables_ = tables;
}

Table Output::only_that_columns(const Table& table, 
                               const Vector<size_t>& only_that_columns) const {
    if (only_that_columns.empty()) {
        return table;
    }
    
    Table new_table;
    for (const auto& row : table) {
        Row new_row;
        for (size_t col_idx : only_that_columns) {
            if (col_idx > 0 && col_idx <= row.size()) {
                new_row.push_back(row[col_idx - 1]);
            }
        }
        if (!new_row.empty()) {
            new_table.push_back(new_row);
        }
    }
    
    if (new_table.empty()) {
        return table;
    }
    return new_table;
}

Vector<std::string> Output::cli_out(const OrderedSet<int32_t>& finally_display_lines,
                                   const Table& new_table,
                                   size_t numlen,
                                   const std::pair<size_t, size_t>& rows_range) {
    Vector<std::string> result;
    
    // Simple implementation for now
    for (size_t row_idx = 0; row_idx < new_table.size(); ++row_idx) {
        std::string line;
        
        if (nummerierung_ && row_idx > 0) {
            char buffer[32];
            snprintf(buffer, sizeof(buffer), "%*zu", (int)numlen, row_idx);
            line += buffer;
            line += " ";
        }
        
        const auto& row = new_table[row_idx];
        for (size_t col_idx = 0; col_idx < row.size(); ++col_idx) {
            if (col_idx > 0) line += " | ";
            
            const auto& cell = row[col_idx];
            if (!cell.lines.empty()) {
                line += cell.lines[0];
            }
        }
        
        result.push_back(line);
    }
    
    resulting_table_ = result;
    return result;
}

// Prepare implementation
Prepare::Prepare() = default;

void Prepare::set_tables(Tables* tables) {
    tables_ = tables;
}

OrderedSet<std::string> Prepare::parameters_cmd_with_some_bereich(
    const std::string& s,
    const std::string& cmd_type,
    const std::string& neg,
    bool keine_neg_beruecksichtigung) const {
    OrderedSet<std::string> result;
    
    auto parts = utils::split(s, ',');
    for (const auto& part : parts) {
        if (part.empty()) continue;
        
        if (cmd_type == "n") {
            auto range = RangeSpec::parse(part);
            auto numbers = range.to_numbers(1000);
            for (auto num : numbers) {
                result.insert(std::to_string(num));
            }
        } else if (cmd_type == "^") {
            try {
                int32_t num = std::stoi(part);
                result.insert(std::to_string(num) + "^2");
            } catch (...) {
                // Ignore invalid numbers
            }
        } else if (cmd_type == "b") {
            try {
                int32_t num = std::stoi(part);
                result.insert(std::to_string(num) + "b");
            } catch (...) {
                // Ignore invalid numbers
            }
        } else {
            result.insert(part);
        }
    }
    
    return result;
}

std::pair<OrderedSet<std::string>, OrderedSet<std::string>> 
Prepare::delete_doubles_in_sets(const OrderedSet<std::string>& set1,
                               const OrderedSet<std::string>& set2) const {
    OrderedSet<std::string> diff1, diff2;
    
    std::set_difference(set1.begin(), set1.end(),
                       set2.begin(), set2.end(),
                       std::inserter(diff1, diff1.begin()));
    
    std::set_difference(set2.begin(), set2.end(),
                       set1.begin(), set1.end(),
                       std::inserter(diff2, diff2.begin()));
    
    return {diff1, diff2};
}

std::tuple<OrderedSet<int32_t>, size_t, Table, size_t, std::pair<size_t, size_t>>
Prepare::prepare4out_before_for_loop_spalten_zeilen_bestimmen(
    const Table& relitable,
    const OrderedSet<std::string>& param_lines,
    const OrderedSet<std::string>& param_lines_not) const {
    OrderedSet<int32_t> finally_display_lines;
    for (int32_t i = 0; i < static_cast<int32_t>(relitable.size()); ++i) {
        finally_display_lines.insert(i);
    }
    
    size_t headings_amount = relitable.empty() ? 0 : relitable[0].size();
    size_t numlen = 3;
    std::pair<size_t, size_t> rows_range = {0, 10};
    
    return {finally_display_lines, headings_amount, relitable, numlen, rows_range};
}

std::tuple<OrderedSet<int32_t>, Table, size_t, 
           std::pair<size_t, size_t>, std::pair<Vector<int32_t>, Vector<int32_t>>>
Prepare::prepare4out(const OrderedSet<std::string>& param_lines,
                   const OrderedSet<std::string>& param_lines_not,
                   const Table& relitable,
                   const OrderedSet<int32_t>& rows_as_numbers,
                   const Map<std::string, OrderedMap<int32_t, Vector<std::string>>>& gebr_spalten,
                   const OrderedMap<int32_t, Vector<std::string>>& prim_spalten) const {
    OrderedSet<int32_t> finally_display_lines;
    for (int32_t i = 0; i < static_cast<int32_t>(relitable.size()); ++i) {
        finally_display_lines.insert(i);
    }
    
    size_t numlen = 3;
    std::pair<size_t, size_t> rows_range = {0, 10};
    std::pair<Vector<int32_t>, Vector<int32_t>> old2new_table;
    
    return {finally_display_lines, relitable, numlen, rows_range, old2new_table};
}

int32_t Prepare::zeile_which_zaehlung(int32_t zeile) const {
    return zeile;
}

Vector<std::string> Prepare::cell_work(const std::string& content, 
                                      int32_t certain_text_width) const {
    if (certain_text_width <= 0) {
        return {content};
    }
    
    Vector<std::string> result;
    std::string current_line;
    size_t width = static_cast<size_t>(certain_text_width);
    
    std::stringstream ss(content);
    std::string word;
    
    while (ss >> word) {
        if (current_line.length() + word.length() + 1 > width && !current_line.empty()) {
            result.push_back(current_line);
            current_line.clear();
        }
        
        if (!current_line.empty()) {
            current_line += " ";
        }
        current_line += word;
    }
    
    if (!current_line.empty()) {
        result.push_back(current_line);
    }
    
    return result;
}

// Combi implementation
Combi::Combi() = default;

void Combi::set_tables(Tables* tables) {
    tables_ = tables;
}

OrderedMap<int32_t, OrderedSet<int32_t>> Combi::prepare_kombi(
    const OrderedSet<int32_t>& finally_display_lines,
    const Table& kombi_table,
    const OrderedSet<std::string>& param_lines,
    const OrderedSet<int32_t>& displaying_zeilen,
    const Vector<Vector<int32_t>>& kombi_table_kombis) const {
    OrderedMap<int32_t, OrderedSet<int32_t>> chosen_kombi_lines;
    
    for (const auto& condition : param_lines) {
        if (condition == "ka" || condition == "ka2") {
            for (size_t kombi_line_number = 0; 
                 kombi_line_number < kombi_table_kombis.size(); 
                 ++kombi_line_number) {
                for (int32_t kombi_number : kombi_table_kombis[kombi_line_number]) {
                    if (displaying_zeilen.find(kombi_number) != displaying_zeilen.end()) {
                        chosen_kombi_lines[kombi_number].insert(
                            static_cast<int32_t>(kombi_line_number) + 1);
                    }
                }
            }
        }
    }
    
    return chosen_kombi_lines;
}

Vector<OrderedMap<int32_t, Table>> Combi::prepare_table_join(
    const OrderedMap<int32_t, OrderedSet<int32_t>>& chosen_kombi_lines,
    const Table& new_table_kombi) const {
    Vector<OrderedMap<int32_t, Table>> kombi_tables;
    
    for (const auto& [key, value] : chosen_kombi_lines) {
        OrderedMap<int32_t, Table> tables_map;
        
        for (int32_t kombi_line_number : value) {
            Set<size_t> lines_allowed = {static_cast<size_t>(kombi_line_number)};
            auto into = tables_->table_reduced_in_lines_by_type_set(new_table_kombi, lines_allowed);
            
            if (!into.empty()) {
                tables_map[key].push_back(into[0]);
            }
        }
        
        if (!tables_map.empty()) {
            kombi_tables.push_back(tables_map);
        }
    }
    
    return kombi_tables;
}

Table Combi::table_join(Table main_table,
                       const Vector<OrderedMap<int32_t, Table>>& many_sub_tables,
                       const std::pair<OrderedMap<int32_t, int32_t>, 
                                       OrderedMap<int32_t, int32_t>>& 
                           maintable2subtable_relation,
                       const std::pair<Vector<int32_t>, Vector<int32_t>>& old2new_rows,
                       const OrderedSet<int32_t>& rows_of_combi) const {
    // Simplified implementation - just return the main table
    return main_table;
}

std::tuple<Table, Table, Vector<Vector<int32_t>>, 
           std::pair<OrderedMap<int32_t, int32_t>, OrderedMap<int32_t, int32_t>>>
Combi::read_kombi_csv(const Table& relitable,
                     OrderedSet<int32_t>& rows_as_numbers,
                     const OrderedSet<int32_t>& rows_of_combi,
                     const std::string& csv_file_name) const {
    // Simplified implementation
    return {Table(), relitable, Vector<Vector<int32_t>>(), 
            {OrderedMap<int32_t, int32_t>(), OrderedMap<int32_t, int32_t>()}};
}

// MainTable implementation
MainTable::MainTable() = default;

void MainTable::set_tables(Tables* tables) {
    tables_ = tables;
}

void MainTable::create_spalte_gestirn(Table& relitable, 
                                     OrderedSet<int32_t>& rows_as_numbers) const {
    if (rows_as_numbers.find(64) != rows_as_numbers.end() && !relitable.empty()) {
        // Add header
        relitable[0].push_back(Cell("Gestirn"));
        
        // Add content for each row
        for (size_t i = 0; i < relitable.size(); ++i) {
            if (i == 0) continue; // Skip header row
            
            std::string content = (i % 2 == 0) ? "Sonne" : "Mond";
            relitable[i].push_back(Cell(content));
        }
        
        // Add to rows_as_numbers
        rows_as_numbers.insert(static_cast<int32_t>(relitable[0].size()) - 1);
    }
}

// Concat implementation
Concat::Concat() = default;

void Concat::set_tables(Tables* tables) {
    tables_ = tables;
}

std::tuple<Table, OrderedSet<int32_t>, OrderedMap<int32_t, Vector<std::string>>>
Concat::read_concat_csv(const Table& relitable,
                       const OrderedSet<int32_t>& rows_as_numbers,
                       const OrderedSet<int32_t>& input,
                       int32_t i) const {
    // Simplified implementation
    return {relitable, rows_as_numbers, OrderedMap<int32_t, Vector<std::string>>()};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat_vervielfache_zeile(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat_modallogik(
    const Table& relitable,
    const OrderedSet<int32_t>& gener_rows,
    const OrderedSet<int32_t>& rows_as_numbers) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat_prim_creativity_type(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat_gleichheit_freiheit_dominieren(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat_geist_emotion_energie_materie_topologie(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat_mond_exponzieren_logarithmus_typ(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat1_row_prim_universe2(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers,
    const OrderedSet<int32_t>& spalten,
    const OrderedMap<std::string, Vector<Vector<std::pair<std::string, std::string>>>>& 
        para_text_namen) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat1_primzahlkreuz_pro_contra(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers,
    const OrderedSet<int32_t>& spalten,
    const i18n::ParametersMain& parameters_main) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::concat_love_polygon(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::spalte_fuer_gegen_innen_aussen_seitlich_prim(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers) const {
    return {relitable, rows_as_numbers};
}

std::pair<Table, OrderedSet<int32_t>> Concat::spalte_meta_kontret_theorie_abstrakt_etc_1(
    const Table& relitable,
    const OrderedSet<int32_t>& rows_as_numbers,
    const Vector<int32_t>& couples_x) const {
    return {relitable, rows_as_numbers};
}

// Tables main implementation
Tables::Tables(std::optional<int32_t> hoechst_zeil, const std::string& txt) 
    : output_(txt) {
    if (hoechst_zeil) {
        config_.hoechste_zeile = {*hoechst_zeil, *hoechst_zeil};
    }
    
    // Set references
    prepare_.set_tables(this);
    output_.set_tables(this);
    combi_.set_tables(this);
    main_table_.set_tables(this);
    concat_.set_tables(this);
}

void Tables::set_hoechste_zeile(int32_t value) {
    config_.hoechste_zeile = {value, value};
}

void Tables::set_text_width(int32_t value) {
    config_.text_width = value;
}

void Tables::set_nummeriere(bool value) {
    config_.nummeriere = value;
}

void Tables::set_breitenn(const Vector<int32_t>& value) {
    config_.breiten = value;
}

Table Tables::table_reduced_in_lines_by_type_set(
    const Table& table,
    const Set<size_t>& lines_allowed) const {
    Table new_table;
    for (size_t i = 0; i < table.size(); ++i) {
        if (lines_allowed.find(i) != lines_allowed.end()) {
            new_table.push_back(table[i]);
        }
    }
    return new_table;
}

std::pair<Vector<std::string>, Vector<std::string>> 
Tables::fill_both(Vector<std::string> liste1, Vector<std::string> liste2) {
    while (liste1.size() < liste2.size()) {
        liste1.push_back("");
    }
    while (liste2.size() < liste1.size()) {
        liste2.push_back("");
    }
    return {liste1, liste2};
}

File: reta2/libs/src3/tables.rs
mod output;
mod combi;
mod prepare;
mod maintable;
mod concat;

use std::collections::{BTreeMap, BTreeSet, HashMap};
use std::sync::Arc;
use crate::types::*;
use crate::error::RetaError;
use crate::i18n;

pub use output::Output;
pub use combi::Combi;
pub use prepare::Prepare;
pub use maintable::MainTable;
pub use concat::Concat;

#[derive(Debug)]
pub struct Tables {
    config: TableConfig,
    color_config: ColorConfig,
    row_num_display_to_orig: OrderedDict<i32, i32>,
    generated_spalten_parameter: OrderedDict<i32, ParameterValue>,
    generated_spalten_parameter_tags: OrderedDict<i32, BTreeSet<SpaltenTag>>,
    prepare: Prepare,
    output: Output,
    combi: Combi,
    main_table: MainTable,
    concat: Concat,
    religion_numbers: Vec<i32>,
    gener_rows: OrderedSet<i32>,
    rows_of_combi: OrderedSet<i32>,
    spalten_vanilla_amount: i32,
    data_dict: Vec<OrderedDict<i32, serde_json::Value>>,
    syntax_type: SyntaxType,
    keine_leeren_inhalte: bool,
    keine_ueberschriften: bool,
    spalte_gestirn: bool,
    if_zeilen_setted: bool,
    if_prim_multis: bool,
    last_line_number: i32,
}

impl Tables {
    pub fn new(hoechst_zeil: Option<i32>, txt: String) -> Self {
        let config = if let Some(hz) = hoechst_zeil {
            TableConfig {
                hoechste_zeile: (hz, hz),
                ..Default::default()
            }
        } else {
            TableConfig::default()
        };
        
        let mut tables = Self {
            config: config.clone(),
            color_config: ColorConfig::default(),
            row_num_display_to_orig: OrderedDict::new(),
            generated_spalten_parameter: OrderedDict::new(),
            generated_spalten_parameter_tags: OrderedDict::new(),
            prepare: Prepare::new(),
            output: Output::new(txt),
            combi: Combi::new(),
            main_table: MainTable::new(),
            concat: Concat::new(),
            religion_numbers: Vec::new(),
            gener_rows: OrderedSet::new(),
            rows_of_combi: OrderedSet::new(),
            spalten_vanilla_amount: 0,
            data_dict: vec![OrderedDict::new(); 14],
            syntax_type: SyntaxType::Default,
            keine_leeren_inhalte: false,
            keine_ueberschriften: false,
            spalte_gestirn: false,
            if_zeilen_setted: false,
            if_prim_multis: false,
            last_line_number: 0,
        };
        
        // Set references
        tables.prepare.set_tables(&tables);
        tables.output.set_tables(&tables);
        tables.combi.set_tables(&tables);
        tables.main_table.set_tables(&tables);
        tables.concat.set_tables(&tables);
        
        tables
    }
    
    // Getters
    pub fn config(&self) -> &TableConfig {
        &self.config
    }
    
    pub fn config_mut(&mut self) -> &mut TableConfig {
        &mut self.config
    }
    
    pub fn prepare(&self) -> &Prepare {
        &self.prepare
    }
    
    pub fn prepare_mut(&mut self) -> &mut Prepare {
        &mut self.prepare
    }
    
    pub fn output(&self) -> &Output {
        &self.output
    }
    
    pub fn output_mut(&mut self) -> &mut Output {
        &mut self.output
    }
    
    pub fn combi(&self) -> &Combi {
        &self.combi
    }
    
    pub fn combi_mut(&mut self) -> &mut Combi {
        &mut self.combi
    }
    
    pub fn concat(&self) -> &Concat {
        &self.concat
    }
    
    pub fn concat_mut(&mut self) -> &mut Concat {
        &mut self.concat
    }
    
    pub fn hoechste_zeile(&self) -> (i32, i32) {
        self.config.hoechste_zeile
    }
    
    pub fn set_hoechste_zeile(&mut self, value: i32) {
        self.config.hoechste_zeile = (value, value);
    }
    
    pub fn text_width(&self) -> i32 {
        self.config.text_width
    }
    
    pub fn set_text_width(&mut self, value: i32) {
        self.config.text_width = value;
    }
    
    pub fn text_height(&self) -> i32 {
        self.config.text_height
    }
    
    pub fn set_text_height(&mut self, value: i32) {
        self.config.text_height = value;
    }
    
    pub fn nummeriere(&self) -> bool {
        self.config.nummeriere
    }
    
    pub fn set_nummeriere(&mut self, value: bool) {
        self.config.nummeriere = value;
    }
    
    pub fn keine_ueberschriften(&self) -> bool {
        self.keine_ueberschriften
    }
    
    pub fn set_keine_ueberschriften(&mut self, value: bool) {
        self.keine_ueberschriften = value;
    }
    
    pub fn keine_leeren_inhalte(&self) -> bool {
        self.keine_leeren_inhalte
    }
    
    pub fn set_keine_leeren_inhalte(&mut self, value: bool) {
        self.keine_leeren_inhalte = value;
    }
    
    pub fn spalte_gestirn(&self) -> bool {
        self.spalte_gestirn
    }
    
    pub fn set_spalte_gestirn(&mut self, value: bool) {
        self.spalte_gestirn = value;
    }
    
    pub fn breitenn(&self) -> &[i32] {
        &self.config.breiten
    }
    
    pub fn set_breitenn(&mut self, value: Vec<i32>) {
        self.config.breiten = value;
    }
    
    pub fn if_zeilen_setted(&self) -> bool {
        self.if_zeilen_setted
    }
    
    pub fn set_if_zeilen_setted(&mut self, value: bool) {
        self.if_zeilen_setted = value;
    }
    
    pub fn if_prim_multis(&self) -> bool {
        self.if_prim_multis
    }
    
    pub fn set_if_prim_multis(&mut self, value: bool) {
        self.if_prim_multis = value;
    }
    
    pub fn gener_rows(&self) -> &OrderedSet<i32> {
        &self.gener_rows
    }
    
    pub fn set_gener_rows(&mut self, value: OrderedSet<i32>) {
        self.gener_rows = value;
    }
    
    pub fn religion_numbers(&self) -> &[i32] {
        &self.religion_numbers
    }
    
    pub fn set_religion_numbers(&mut self, value: Vec<i32>) {
        self.religion_numbers = value;
    }
    
    pub fn rows_of_combi(&self) -> &OrderedSet<i32> {
        &self.rows_of_combi
    }
    
    pub fn set_rows_of_combi(&mut self, value: OrderedSet<i32>) {
        self.rows_of_combi = value;
    }
    
    pub fn spalten_vanilla_amount(&self) -> i32 {
        self.spalten_vanilla_amount
    }
    
    pub fn set_spalten_vanilla_amount(&mut self, value: i32) {
        self.spalten_vanilla_amount = value;
    }
    
    pub fn data_dict(&self) -> &[OrderedDict<i32, serde_json::Value>] {
        &self.data_dict
    }
    
    pub fn set_data_dict(&mut self, value: Vec<OrderedDict<i32, serde_json::Value>>) {
        self.data_dict = value;
    }
    
    pub fn syntax_type(&self) -> SyntaxType {
        self.syntax_type
    }
    
    pub fn set_syntax_type(&mut self, value: SyntaxType) {
        self.syntax_type = value;
    }
    
    pub fn last_line_number(&self) -> i32 {
        self.last_line_number
    }
    
    pub fn set_last_line_number(&mut self, value: i32) {
        self.last_line_number = value;
    }
    
    pub fn table_reduced_in_lines_by_type_set(
        &self,
        table: &Table,
        lines_allowed: &BTreeSet<usize>,
    ) -> Table {
        let mut new_table = Table::new();
        for (i, row) in table.iter().enumerate() {
            if lines_allowed.contains(&i) {
                new_table.push(row.clone());
            }
        }
        new_table
    }
    
    pub fn fill_both(mut liste1: Vec<String>, mut liste2: Vec<String>) -> (Vec<String>, Vec<String>) {
        while liste1.len() < liste2.len() {
            liste1.push(String::new());
        }
        while liste2.len() < liste1.len() {
            liste2.push(String::new());
        }
        (liste1, liste2)
    }
}

// Output submodule
pub mod output {
    use super::*;
    use crate::utils;
    
    #[derive(Debug)]
    pub struct Output {
        txt: String,
        resulting_table: Vec<String>,
        color: bool,
        one_table: bool,
        breiten: Vec<i32>,
        nummerierung: bool,
        textheight: i32,
        textwidth: i32,
        religion_numbers: Vec<i32>,
        rows_as_numbers: OrderedSet<i32>,
        tables: Option<*const Tables>, // Raw pointer for self-reference
    }
    
    impl Output {
        pub fn new(txt: String) -> Self {
            Self {
                txt,
                resulting_table: Vec::new(),
                color: true,
                one_table: false,
                breiten: Vec::new(),
                nummerierung: true,
                textheight: 0,
                textwidth: 21,
                religion_numbers: Vec::new(),
                rows_as_numbers: OrderedSet::new(),
                tables: None,
            }
        }
        
        pub fn set_tables(&mut self, tables: &Tables) {
            self.tables = Some(tables as *const Tables);
        }
        
        fn tables(&self) -> &Tables {
            unsafe { &*self.tables.unwrap() }
        }
        
        pub fn color(&self) -> bool {
            self.color
        }
        
        pub fn set_color(&mut self, value: bool) {
            self.color = value;
        }
        
        pub fn one_table(&self) -> bool {
            self.one_table
        }
        
        pub fn set_one_table(&mut self, value: bool) {
            self.one_table = value;
        }
        
        pub fn rows_as_numbers(&self) -> &OrderedSet<i32> {
            &self.rows_as_numbers
        }
        
        pub fn set_rows_as_numbers(&mut self, value: OrderedSet<i32>) {
            self.rows_as_numbers = value;
        }
        
        pub fn only_that_columns(&self, table: &Table, only_that_columns: &[usize]) -> Table {
            if only_that_columns.is_empty() {
                return table.clone();
            }
            
            let mut new_table = Table::new();
            for row in table {
                let mut new_row = Row::new();
                for &col_idx in only_that_columns {
                    if col_idx > 0 && col_idx <= row.len() {
                        new_row.push(row[col_idx - 1].clone());
                    }
                }
                if !new_row.is_empty() {
                    new_table.push(new_row);
                }
            }
            
            if new_table.is_empty() {
                table.clone()
            } else {
                new_table
            }
        }
        
        pub fn cli_out(
            &mut self,
            finally_display_lines: &BTreeSet<i32>,
            new_table: &Table,
            numlen: usize,
            rows_range: std::ops::Range<usize>,
        ) -> Vec<String> {
            // Implementation of complex output logic
            let mut result = Vec::new();
            
            // Simple implementation for now
            for (row_idx, row) in new_table.iter().enumerate() {
                let mut line_parts = Vec::new();
                
                if self.nummerierung && row_idx > 0 {
                    line_parts.push(format!("{:>numlen$}", row_idx));
                }
                
                for (col_idx, cell) in row.iter().enumerate() {
                    let content = if !cell.lines.is_empty() {
                        cell.lines[0].clone()
                    } else {
                        "".to_string()
                    };
                    line_parts.push(content);
                }
                
                result.push(line_parts.join(" | "));
            }
            
            self.resulting_table = result.clone();
            result
        }
    }
}

// Prepare submodule
pub mod prepare {
    use super::*;
    
    #[derive(Debug)]
    pub struct Prepare {
        ifprimmultis: bool,
        if_zeilen_setted: bool,
        breiten: Vec<i32>,
        nummerierung: bool,
        text_width: i32,
        religion_numbers: Vec<i32>,
        rows_as_numbers: OrderedSet<i32>,
        tables: Option<*const Tables>,
    }
    
    impl Prepare {
        pub fn new() -> Self {
            Self {
                ifprimmultis: false,
                if_zeilen_setted: false,
                breiten: Vec::new(),
                nummerierung: true,
                text_width: 0,
                religion_numbers: Vec::new(),
                rows_as_numbers: OrderedSet::new(),
                tables: None,
            }
        }
        
        pub fn set_tables(&mut self, tables: &Tables) {
            self.tables = Some(tables as *const Tables);
        }
        
        pub fn ifprimmultis(&self) -> bool {
            self.ifprimmultis
        }
        
        pub fn set_ifprimmultis(&mut self, value: bool) {
            self.ifprimmultis = value;
        }
        
        pub fn if_zeilen_setted(&self) -> bool {
            self.if_zeilen_setted
        }
        
        pub fn set_if_zeilen_setted(&mut self, value: bool) {
            self.if_zeilen_setted = value;
        }
        
        pub fn rows_as_numbers(&self) -> &OrderedSet<i32> {
            &self.rows_as_numbers
        }
        
        pub fn set_rows_as_numbers(&mut self, value: OrderedSet<i32>) {
            self.rows_as_numbers = value;
        }
        
        pub fn parameters_cmd_with_some_bereich(
            &self,
            s: &str,
            cmd_type: &str,
            neg: &str,
            keine_neg_beruecksichtigung: bool,
        ) -> OrderedSet<String> {
            let mut result = OrderedSet::new();
            
            // Parse range specifications
            for part in s.split(',') {
                let part = part.trim();
                if part.is_empty() {
                    continue;
                }
                
                match cmd_type {
                    "n" => {
                        // Number range
                        if let Ok(range) = RangeSpec::parse(part) {
                            let numbers = range.to_numbers(1000); // Default max
                            for num in numbers {
                                result.insert(num.to_string());
                            }
                        }
                    }
                    "^" => {
                        // Powers
                        if let Ok(num) = part.parse::<i32>() {
                            result.insert(format!("{}^{}", num, 2)); // Default power 2
                        }
                    }
                    "b" => {
                        // Multiples
                        if let Ok(num) = part.parse::<i32>() {
                            result.insert(format!("{}b", num));
                        }
                    }
                    _ => {
                        result.insert(part.to_string());
                    }
                }
            }
            
            result
        }
        
        pub fn delete_doubles_in_sets(
            &self,
            set1: OrderedSet<String>,
            set2: OrderedSet<String>,
        ) -> (OrderedSet<String>, OrderedSet<String>) {
            let diff1: OrderedSet<String> = set1.difference(&set2).cloned().collect();
            let diff2: OrderedSet<String> = set2.difference(&set1).cloned().collect();
            (diff1, diff2)
        }
        
        pub fn prepare4out_before_for_loop_spalten_zeilen_bestimmen(
            &self,
            relitable: &Table,
            param_lines: &OrderedSet<String>,
            param_lines_not: &OrderedSet<String>,
        ) -> (OrderedSet<i32>, usize, Table, usize, std::ops::Range<usize>) {
            // Simplified implementation
            let finally_display_lines: OrderedSet<i32> = (0..relitable.len() as i32).collect();
            let headings_amount = if !relitable.is_empty() { relitable[0].len() } else { 0 };
            let newer_table = relitable.clone();
            let numlen = 3; // Default
            let rows_range = 0..10; // Default range
            
            (finally_display_lines, headings_amount, newer_table, numlen, rows_range)
        }
        
        pub fn prepare4out(
            &self,
            param_lines: &OrderedSet<String>,
            param_lines_not: &OrderedSet<String>,
            relitable: &Table,
            rows_as_numbers: &OrderedSet<i32>,
            gebr_spalten: &HashMap<String, OrderedDict<i32, Vec<String>>>,
            prim_spalten: &OrderedDict<i32, Vec<String>>,
        ) -> (
            OrderedSet<i32>,
            Table,
            usize,
            std::ops::Range<usize>,
            (Vec<i32>, Vec<i32>),
        ) {
            // Simplified implementation
            let finally_display_lines: OrderedSet<i32> = (0..relitable.len() as i32).collect();
            let table = relitable.clone();
            let numlen = 3;
            let rows_range = 0..10;
            let old2new_table = (Vec::new(), Vec::new());
            
            (finally_display_lines, table, numlen, rows_range, old2new_table)
        }
        
        pub fn zeile_which_zaehlung(&self, zeile: i32) -> i32 {
            // Simple implementation - just return the line number
            zeile
        }
        
        pub fn cell_work(&self, content: &str, certain_text_width: i32) -> Vec<String> {
            let width = certain_text_width as usize;
            let mut result = Vec::new();
            
            if width == 0 {
                return vec![content.to_string()];
            }
            
            let mut current_line = String::new();
            for word in content.split_whitespace() {
                if current_line.len() + word.len() + 1 > width && !current_line.is_empty() {
                    result.push(current_line);
                    current_line = String::new();
                }
                
                if !current_line.is_empty() {
                    current_line.push(' ');
                }
                current_line.push_str(word);
            }
            
            if !current_line.is_empty() {
                result.push(current_line);
            }
            
            result
        }
    }
}

// Combi submodule
pub mod combi {
    use super::*;
    use std::fs::File;
    use std::io::{BufRead, BufReader};
    
    #[derive(Debug)]
    pub struct Combi {
        sum_of_all_combi_rows_amount: i32,
        religion_numbers: Vec<i32>,
        rows_of_combi: OrderedSet<i32>,
        tables: Option<*const Tables>,
    }
    
    impl Combi {
        pub fn new() -> Self {
            Self {
                sum_of_all_combi_rows_amount: 0,
                religion_numbers: Vec::new(),
                rows_of_combi: OrderedSet::new(),
                tables: None,
            }
        }
        
        pub fn set_tables(&mut self, tables: &Tables) {
            self.tables = Some(tables as *const Tables);
        }
        
        pub fn sum_of_all_combi_rows_amount(&self) -> i32 {
            self.sum_of_all_combi_rows_amount
        }
        
        pub fn prepare_kombi(
            &self,
            finally_display_lines: &BTreeSet<i32>,
            kombi_table: &Table,
            param_lines: &BTreeSet<String>,
            displaying_zeilen: &BTreeSet<i32>,
            kombi_table_kombis: &[Vec<i32>],
        ) -> OrderedDict<i32, OrderedSet<i32>> {
            let mut chosen_kombi_lines = OrderedDict::
